source-test/File.ts

```diff
@@ -1,661 +1,1617 @@
+        cwd: '/',
+        base: '/test/',
+        path: '/test/test.coffee',
+        contents: from(['wa', 'dup']),
+      };
+      const file = new Vinyl(options);
+      const file2 = file.clone();
+      let ends = 2;
+
+      let data = '';
+      let data2 = '';
+
+      function assert() {
+        if (--ends === 0) {
+          expect(data).toEqual(data2);
+          done();
+        }
+      }
+
+      // Start flowing file2
+      (<NodeJS.ReadableStream> file2.contents).on('data', (chunk: Buffer) => {
+        data2 += chunk.toString('utf8');
       });
-      stream.on('end', () => {
-        done();
+
+      process.nextTick(function() {
+        // Nothing was written yet
+        expect(data).toEqual('');
+        expect(data2).toEqual('');
+
+        // Starts flowing file
+        (<NodeJS.ReadableStream> file.contents).on('data', (chunk: Buffer) => {
+          data += chunk.toString('utf8');
+        });
       });
-      const ret = file.pipe(stream);
-      ret.should.equal(stream, 'should return the stream');
+
+      (<NodeJS.ReadableStream> file2.contents).on('end', assert);
+      (<NodeJS.ReadableStream> file.contents).on('end', assert);
     });
 
-    it('should pipe to stream with Stream', done => {
-      const testChunk = new Buffer('test');
+    it('does not start flowing until all clones flows (readable)', (done) => {
       const options = {
         cwd: '/',
         base: '/test/',
         path: '/test/test.coffee',
-        contents: new Stream.PassThrough()
+        contents: from(['wa', 'dup']),
       };
-      const file = new File(options);
-      const stream = new Stream.PassThrough();
-      stream.on('data', (chunk: any) => {
-        should.exist(chunk);
-        (chunk instanceof Buffer).should.equal(true, 'should write as a buffer');
-        chunk.toString('utf8').should.equal(testChunk.toString('utf8'));
-        done();
-      });
-      const ret = file.pipe(stream);
-      ret.should.equal(stream, 'should return the stream');
-
-      let fileContents = file.contents;
-      if (fileContents instanceof Buffer) {
-        fileContents.write(testChunk.toString());
+      const file = new Vinyl(options);
+      const file2 = file.clone();
+
+      let data2 = '';
+
+      function assert(data: Buffer) {
+        expect(data.toString('utf8')).toEqual(data2);
       }
+
+      // Start flowing file2
+      (<NodeJS.ReadableStream> file2.contents).on('readable', () => {
+        let chunk: Buffer;
+        while ((chunk = this.read()) !== null) {
+          data2 += chunk.toString();
+        }
+      });
+
+      pipe([
+        file.contents,
+        concat(assert),
+      ], done);
     });
 
-    it('should do nothing with null', done => {
+    it('copies all attributes over with null contents', function(done) {
       const options = {
         cwd: '/',
         base: '/test/',
         path: '/test/test.coffee',
-        contents: fakeStream
+        contents: <Buffer> null,
```

Comment:
I guess it's related to the above: the compiler does not know if it is a Buffer or a ReadableSteam.